README.rst
.. -*- mode: rst; coding: utf-8 -*-

.. image:: pypiserver_logo.png
   :width: 300 px
   :align: center

==============================================================================
pypiserver - minimal PyPI server for use with pip/easy_install
==============================================================================
|pypi-ver| |travis-status| |dependencies| |downloads-count| |python-ver| \
|proj-license|

:Version:     1.2.0.dev2
:Date:        2016-XX-XX
:Source:      https://github.com/pypiserver/pypiserver
:PyPI:        https://pypi.python.org/pypi/pypiserver
:Travis:      https://travis-ci.org/pypiserver/pypiserver
:Maintainer:  Kostis Anagnostopoulos <ankostis@gmail.com>
:License:     zlib/libpng + MIT

*pypiserver* is a minimal PyPI_ compatible server for *pip* or *easy_install*.
It is based on bottle_ and serves packages from regular directories.
Wheels, bdists, eggs and accompanying PGP-signatures can be uploaded
either with *pip*, *setuptools*, *twine* or simply copied with *scp*.


.. contents:: Table of Contents
  :backlinks: top


Quickstart: Installation and Usage
==================================
*pypiserver* `> 1.2.x` works with python `2.7` and `3.3 --> 3.5` or *pypy*.
Python `3.0 --> 3.2` may also work, but it is not being tested for these
versions.
For legacy python versions, use `pypiserver-1.1.x` series.

Run the following commands to get your PyPI server up and running::

  ## Installation.
  pip install pypiserver                ## Or: pypiserver[passlib,watchdog]
  mkdir ~/packages                      ## Copy packages into this directory.

  ## Start server.
  pypi-server -p 8080 ~/packages &      ## Will listen to all IPs.

From the client computer, type this::

  ## Download and Install hosted packages.
  pip install  --extra-index-url http://localhost:8080/simple/ ...
  ## Search hosted packages
  pip search --index http://localhost:8080/simple/ ...

See also `Client-side configurations`_ for avoiding tedious typing.

.. Note::
  The above commands work on a unix-like operating system with a posix shell.
  The ``~`` character expands to user's home directory.

  If you're using windows, you'll have to use their "windows counterparts".
  The same is true for the rest of this documentation.


Uploading packages from sources, remotely
-----------------------------------------
Instead of copying packages directly to the server's folder,
you may also upload them remotely with a ``python setup.py upload`` command.
Currently only password-protected uploads are supported!

#. First make sure you have the *passlib* module installed (note that
   `passlib>=1.6` is required), which is needed for parsing the apache
   *htpasswd* file specified by the `-P`, `--passwords` option
   (see next steps)::

     pip install passlib

#. Create the apache *htpasswd* file with at least one user/password pair
   with this command (you'll be prompted for a password)::

     htpasswd -sc htpasswd.txt <some_username>

   .. Tip::
     Read this SO question for running `htpasswd` cmd under *Windows*:

        http://serverfault.com/questions/152950/how-to-create-and-edit-htaccess-and-htpasswd-locally-on-my-computer-and-then-u

     or if you have bogus passwords for an internal service you may use this
     public service:

        http://www.htaccesstools.com/htpasswd-generator/

     It is also possible to disable authentication even for uploads.
     To avoid lazy security decisions, read help for ``-P`` and ``-a`` options.

#. You  need to restart the server with the `-P` option only once
   (but user/password pairs can later be added or updated on the fly)::

     ./pypi-server -p 8080 -P htpasswd.txt ~/packages &

#. On client-side, edit or create a `~/.pypirc` file with a similar content::

     [distutils]
     index-servers =
       pypi
       local

     [pypi]
     username:<your_pypi_username>
     password:<your_pypi_passwd>

     [local]
     repository: http://localhost:8080
     username: <some_username>
     password: <some_passwd>

#. Then from within the directory of the python-project you wish to upload,
   issue this command::

     python setup.py sdist upload -r local

.. Tip::
    To avoid storing you passwords on disk in clear text, you may either:
       - Use the `register` command with the `-r` option, like that::

           python setup.py sdist register -r local upload -r local

       - Use `twine <https://pypi.python.org/pypi/twine>`_ library which
         breaks the procedure in two steps.

         Furthermore, it allows you to pre-sign your files with PGP-Signatures
         and upload also the generated `.asc` files to *pypiserver*::

            twine upload -r local --sign -identity user_name ./foo-1.zip



Client-side configurations
--------------------------
Always specifying the the pypi url on the command line is a bit
cumbersome. Since pypi-server redirects pip/easy_install to the
pypi.python.org index if it doesn't have a requested package, it's a
good idea to configure them to always use your local pypi index.

`pip`
~~~~~
For *pip* this can be done by setting the environment variable
`PIP_EXTRA_INDEX_URL` in your `.bashrc`/`.profile`/`.zshrc`::

  export PIP_EXTRA_INDEX_URL=http://localhost:8080/simple/

or by adding the following lines to `~/.pip/pip.conf`::

  [global]
  extra-index-url = http://localhost:8080/simple/

.. Note::
    If you have installed *pypi-server* on a remote url without *https*
    you wil receive an "untrusted" warning from *pip*, urging you to append
    the `--trusted-host` option.  You can also include this option permanently
    in your configuration-files or environment variables.


`easy_install`
~~~~~~~~~~~~~~
For *easy_install* it can be configured with the following setting in
`~/.pydistutils.cfg`::

  [easy_install]
  index_url = http://localhost:8080/simple/


Alternative Installation methods
================================
When trying the methods below, first use the following command to check whether
previous versions of *pypiserver* already exist, and (optionally) uninstall them::

  ## VERSION-CHECK: Fails if not installed.
  pypi-server --version

  ## UNINSTALL: Invoke again untill it fails.
  pip uninstall pypiserver


Installing the very latest version
----------------------------------
In case the latest version in *pypi* is a pre-release, you have to use
*pip*'s `--pre` option.  And to update an existing installation combine it
with `--ignore-installed`::

  pip install pypiserver --pre -I

You can even install the latest *pypiserver* directly from *github* with the
following command, assuming you have *git* installed on your `$PATH`::

  pip install git+git://github.com/pypiserver/pypiserver.git


Installing it as standalone script
----------------------------------
The git repository contains a ``pypi-server-standalone.py`` script,
which is a single python file that can be executed without any other
dependencies.

Run the following commands to download the script with `wget`::

  wget https://raw.github.com/pypiserver/pypiserver/standalone/pypi-server-standalone.py
  chmod +x pypi-server-standalone.py

or with `curl`::

  curl -O https://raw.github.com/pypiserver/pypiserver/standalone/pypi-server-standalone.py
  chmod +x pypi-server-standalone.py

You can then start-up the server with::

  ./pypi-server-standalone.py

Feel free to rename the script and move it into your `$PATH`.


Running on heroku/dotcloud
--------------------------
https://github.com/dexterous/pypiserver-on-the-cloud contains
instructions on how to run pypiserver on one of the supported cloud
service providers.

Running as *systemd* service
----------------------------
A template for *systemd* is given that must be adapted and written into
``/usr/lib/systemd/system/``::

    # /lib/systemd/system/pypi-server.service
    [Unit]
    Description=pypi-server
    After=network.target

    [Service]
    Type=simple
    # systemd requires absolute path here too.
    PIDFile=/var/run/pypi-server.pid
    User=www-data
    Group=www-data

    ExecStart=/usr/local/bin/pypi-server -p 8080 -vvv -u --disable-fallback --authenticate update,download --log-file /var/log/pypi-server.log --passwords /etc/nginx/.htpasswd /var/www/pypi
    ExecStop=/bin/kill -TERM $MAINPID
    ExecReload=/bin/kill -HUP $MAINPID
    Restart=always

    WorkingDirectory=/var/www/pypi

    TimeoutStartSec=3
    RestartSec=5

    [Install]
    WantedBy=multi-user.target

More information about *systemd* can be found at
https://www.digitalocean.com/community/tutorials/how-to-use-systemctl-to-manage-systemd-services-and-units


Detailed Usage
=================================
Running ``pypi-server -h`` will print a detailed usage message::

  pypi-server [OPTIONS] [PACKAGES_DIRECTORY...]
    start PyPI compatible package server serving packages from
    PACKAGES_DIRECTORY. If PACKAGES_DIRECTORY is not given on the
    command line, it uses the default ~/packages.  pypiserver scans this
    directory recursively for packages. It skips packages and
    directories starting with a dot. Multiple package directories can be
    specified.

  pypi-server understands the following options:

    -p, --port PORT
      listen on port PORT (default: 8080)

    -i, --interface INTERFACE
      listen on interface INTERFACE (default: 0.0.0.0, any interface)

    -a, --authenticate (UPDATE|download|list), ...
      comma-separated list of (case-insensitive) actions to authenticate
      Use '.' or '' for empty. Requires to have set the password (-P option).
      For example to password-protect package downloads (in addition to uploads)
      while leaving listings public, give:
        -P foo/htpasswd.txt  -a update,download
      To drop all authentications, use:
        -P .  -a .
      Note that when uploads are not protected, the `register` command
      is not necessary, but `~/.pypirc` still need username and password fields,
      even if bogus.
      By default, only 'update' is password-protected.

    -P, --passwords PASSWORD_FILE
      use apache htpasswd file PASSWORD_FILE to set usernames & passwords when
      authenticating certain actions (see -a option).
      If you want to allow un-authorized access, set this option and -a
      explicitly to empty (either '.' or'').

    --disable-fallback
      disable redirect to real PyPI index for packages not found in the
      local index

    --fallback-url FALLBACK_URL
      for packages not found in the local index, this URL will be used to
      redirect to (default: http://pypi.python.org/simple)

    --server METHOD
      use METHOD to run the server. Valid values include paste,
      cherrypy, twisted, gunicorn, gevent, wsgiref, auto. The
      default is to use "auto" which chooses one of paste, cherrypy,
      twisted or wsgiref.

    -r, --root PACKAGES_DIRECTORY
      [deprecated] serve packages from PACKAGES_DIRECTORY

    -o, --overwrite
      allow overwriting existing package files

    --hash-algo ALGO
      any `hashlib` available algo used as fragments on package links.
      Set one of (0, no, off, false) to disabled it. (default: md5)

    --welcome HTML_FILE
      uses the ASCII contents of HTML_FILE as welcome message response.

    -v
      enable INFO logging;  repeat for more verbosity.

    --log-conf <FILE>
      read logging configuration from FILE.
      By default, configuration is read from `log.conf` if found in server's dir.

    --log-file <FILE>
      write logging info into this FILE.

    --log-frmt <FILE>
      the logging format-string.  (see `logging.LogRecord` class from standard python library)
      [Default: %(asctime)s|%(name)s|%(levelname)s|%(thread)d|%(message)s]

    --log-req-frmt FORMAT
      a format-string selecting Http-Request properties to log; set to  '%s' to see them all.
      [Default: %(bottle.request)s]

    --log-res-frmt FORMAT
      a format-string selecting Http-Response properties to log; set to  '%s' to see them all.
      [Default: %(status)s]

    --log-err-frmt FORMAT
      a format-string selecting Http-Error properties to log; set to  '%s' to see them all.
      [Default: %(body)s: %(exception)s \n%(traceback)s]

  pypi-server -h
  pypi-server --help
    show this help message

  pypi-server --version
    show pypi-server's version

  pypi-server -U [OPTIONS] [PACKAGES_DIRECTORY...]
    update packages in PACKAGES_DIRECTORY. This command searches
    pypi.python.org for updates and shows a pip command line which
    updates the package.

  The following additional options can be specified with -U:

    -x
      execute the pip commands instead of only showing them

    -d DOWNLOAD_DIRECTORY
      download package updates to this directory. The default is to use
      the directory which contains the latest version of the package to
      be updated.

    -u
      allow updating to unstable version (alpha, beta, rc, dev versions)

  Visit https://github.com/pypiserver/pypiserver for more information.



Managing the package directory
------------------------------
The `pypi-server` command has the `-U` option that searches for updates of
available packages. It scans the package directory for available
packages and searches on pypi.python.org for updates. Without further
options ``pypi-server -U`` will just print a list of commands which must
be run in order to get the latest version of each package. Output
looks like::

   $ ./pypi-server -U
  checking 106 packages for newer version

  .........u.e...........e..u.............
  .....e..............................e...
  ..........................

  no releases found on pypi for PyXML, Pymacs, mercurial, setuptools

  # update raven from 1.4.3 to 1.4.4
  pip -q install --no-deps  --extra-index-url http://pypi.python.org/simple -d /home/ralf/packages/mirror raven==1.4.4

  # update greenlet from 0.3.3 to 0.3.4
  pip -q install --no-deps  --extra-index-url http://pypi.python.org/simple -d /home/ralf/packages/mirror greenlet==0.3.4

It first prints for each package a single character after checking the
available versions on pypi. A dot(`.`) means the package is up-to-date, `u`
means the package can be updated and `e` means the list of releases on
pypi is empty. After that it shows a *pip* command line which can be used
to update a one package. Either copy and paste that or run
``pypi-server -Ux`` in order to really execute those commands. You need
to have *pip* installed for that to work however.

Specifying an additional `-u` option will also allow alpha, beta and
release candidates to be downloaded. Without this option these
releases won't be considered.


Serving thousands of packages
-----------------------------

By default, *pypiserver* scans the entire packages directory each time an
incoming HTTP request occurs. This isn't a problem for a small number of
packages, but causes noticeable slowdowns when serving thousands or tens
of thousands of packages.

If you run into this problem, significant speedups can be gained by enabling
pypiserver's directory caching functionality. The only requirement is to
install the ``watchdog`` package, or it can be installed by installing
``pypiserver`` using the ``cache`` extras option::

  pip install pypiserver[cache]

If you are using a static webserver such as *Apache* or *nginx* as
a reverse-proxy for pypiserver, additional speedup can be gained by
directly serving the packages directory:

For instance, in *nginx* you may adding the following config to serve
packages-directly directly (take care not to expose "sensitive" files)::

    location /packages/ {
      root /path/to/packages/parentdir;
    }

If you have packages that are very large, you may find it helpful to
disable hashing of files (set `--hash-algo=off`, or `hash_algo=None` when
using wsgi).

Using a different WSGI server
-----------------------------
- *pypiserver* ships with it's own copy of bottle.
  It's possible to use bottle with different WSGI servers.

- *pypiserver* chooses any of the
  following *paste*, *cherrypy*, *twisted*, *wsgiref* (part of python) if
  available.

- If none of the above servers matches your needs, pypiserver also
  exposes an API to get the internal WSGI app, which you can then run
  under any WSGI server you like. `pypiserver.app` has the following
  interface::

    def app(root=None,
        redirect_to_fallback=True,
        fallback_url="http://pypi.python.org/simple")

  and returns the WSGI application. `root` is the package directory,
  `redirect_to_fallback` specifies whether to redirect to `fallback_url` when
  a package is missing.


gunicorn
~~~~~~~~

The following command uses *gunicorn* to start *pypiserver*::

  gunicorn -w4 'pypiserver:app(root="/home/ralf/packages")'

or when using multiple roots::

  gunicorn -w4 'pypiserver:app(root=["/home/ralf/packages", "/home/ralf/experimental"])'


apache/mod_wsgi
~~~~~~~~~~~~~~~
In case you're using *apache2* with *mod_wsgi*, the following config-file
(contributed by Thomas Waldmann) can be used::

  # An example pypiserver.wsgi for use with apache2 and mod_wsgi, edit as necessary.
  #
  # apache virtualhost configuration for mod_wsgi daemon mode:
  #    Alias /robots.txt /srv/yoursite/htdocs/robots.txt
  #    WSGIPassAuthorization On
  #    WSGIScriptAlias /     /srv/yoursite/cfg/pypiserver.wsgi
  #    WSGIDaemonProcess     pypisrv user=pypisrv group=pypisrv processes=1 threads=5 maximum-requests=500 umask=0007 display-name=wsgi-pypisrv inactivity-timeout=300
  #    WSGIProcessGroup      pypisrv

  PACKAGES = "/srv/yoursite/packages"
  HTPASSWD = "/srv/yoursite/htpasswd"
  import pypiserver
  application = pypiserver.app(root=PACKAGES, redirect_to_fallback=True, password_file=HTPASSWD)


paste/pastedeploy
~~~~~~~~~~~~~~~~~
*paste* allows to run multiple WSGI applications under different URL
paths. Therefore it's possible to serve different set of packages on
different paths.

The following example `paste.ini` could be used to serve stable and
unstable packages on different paths::

  [composite:main]
  use = egg:Paste#urlmap
  /unstable/ = unstable
  / = stable

  [app:stable]
  use = egg:pypiserver#main
  root = ~/stable-packages

  [app:unstable]
  use = egg:pypiserver#main
  root = ~/stable-packages
     ~/unstable-packages

  [server:main]
  use = egg:gunicorn#main
  host = 0.0.0.0
  port = 9000
  workers = 5
  accesslog = -

.. Note::
  You need to install some more dependencies for this to work,
  e.g. run::

    pip install paste pastedeploy gunicorn pypiserver

  The server can then be started with::

    gunicorn_paster paste.ini



Sources
=======
Use::

  git clone https://github.com/pypiserver/pypiserver.git

to create a copy of the repository, then::

  git pull

inside the copy to receive any later changes.



Known Limitations
=================
*pypiserver* does not implement the full API as seen on PyPI_. It
implements just enough to make ``easy_install`` and ``pip install`` to work.

The following limitations are known:

- It doesn't implement the XMLRPC json API interface: pip search
  will not work (a patch has been suggested:
  issue https://github.com/pypiserver/pypiserver/issues/80).
- Command ``pypi -U`` that compares uploaded packages with *pypi* to see if
  they are outdated does not respect a http-proxy environment variable
  (see https://github.com/pypiserver/pypiserver/issues/19).
- It accepts documentation uploads but does not save them to
  disk (see https://github.com/pypiserver/pypiserver/issues/47 for a
  discussion)
- It does not handle misspelled packages as *pypi-repo* does,
  therefore it is suggested to use it with `--extra-index-url` instead
  of `--index-url` (see https://github.com/pypiserver/pypiserver/issues/38)

Please use github's `bugtracker <https://github.com/pypiserver/pypiserver/issues>`_
for other bugs you find.



Similar Projects
================
There are lots of other projects, which allow you to run your own
PyPI server. If *pypiserver* doesn't work for you, the following are
among the most popular alternatives:

- `devpi-server <https://pypi.python.org/pypi/devpi-server>`_:
  a reliable fast pypi.python.org caching server, part of
  the comprehensive `github-style pypi index server and packaging meta tool
  <https://pypi.python.org/pypi/devpi>`_.
  (version: 2.1.4, access date: 8/3/2015)

- `pypi-uploader <https://pypi.python.org/pypi/pypi-uploader>`_ -
  Cmd-line tool to download and re-uploads distributions or your requirements
  to *pypiserver*.

- `pip2pi <https://github.com/wolever/pip2pi>`_ -
  a simple cmd-line tool that builds a PyPI-compatible local folder from pip requirements
  (version: 0.6.7, access date: 8/3/2015)

- `flask-pypi-proxy <http://flask-pypi-proxy.readthedocs.org/>`_ -
  A proxy for PyPI that also enables also uploading custom packages.

-
- Check this SO question: ` How to roll my own pypi <http://stackoverflow.com/questions/1235331/how-to-roll-my-own-pypi>`_



License
=======
*pypiserver* contains a copy of bottle_ which is available under the
*MIT* license, and the remaining part is distributed under the *zlib/libpng* license.
See the `LICENSE.txt` file.



.. _bottle: http://bottlepy.org
.. _PyPI: http://pypi.python.org

CHANGES.rst
Changelog
=========

1.2.0 (2016-XX-XX)
------------------
"New Day": less rigorous support for ``python-2 < 2.7`` and ``python-3 < 3.3``.

- #38, #138, #139, #140: Package normalizations and :pep:`503` updates
  (thanks to @dpkp):

  - URLs are redirected unless they end in `'/'` (expect packages themselves).
  - Package names are normalized: lower-case with all runs of the characters,
    and replace `[-_.]` with a single `'-'` character.
  - The simple index only lists normalized package names.
  - Any request for a non-normalized package name is redirected to
    the normalized name.

- #117, #122, #124/#127/#128: FIX startup regressions introduced by
  previous `v1.1.10` (thanks to @virtuald, @Oneplus, @michaelkuty, @harcher81,
  @8u1a).
- #113: FIX over-writing of packages even when without `--overwrite` flag
  (thanks @blade2005).
- #107, #108, #110: Test against *python-3.5*.
- #112, #118, #119: Fixes for *paste*, *gunicorn* and other *WSGI* servers
  (thanks to @corywright, @virtuald, @montefra).
- #120/#121, #129, #141/#142: Updates and fixes needed due to changes
  in dependent libraries (thanks @dpkp).
- #116: Add cache for speeding up GPG signatures (thanks to @virtuald).
- Provide sample ``systemd`` *unit-file* (thanks to @ssbarnea).
- Add usage instructions for related project ``pypi-uploader`` and mention also
  project
  (thanks @ssbarnea & @bibby).
- #129, #131: Other minor fixes and improvements (thanks to @bibby, @Oneplus,
  @8u1a).
- Detailed changes recorded in `Github's milestone 1.2.0
  <https://github.com/pypiserver/pypiserver/milestones/M1.2.0>`_.


1.1.10 (2016-01-19)
-------------------
Serve 1000s of packages, PGP-Sigs, skip versions starting with 'v'.

+ #101: Speed-up server by (optionally) using the `watchdog` package
  to cache results, serve packages directly from proxying-server (*Apache* ,
  *nginx*), and pre-compile regexes (thanks @virtuald).
- #106: Support uploading PGP-signatures (thanks @mplanchard).
- Package-versions parsing modifications:

  - #104: Stopped parsing invalid package-versions prefixed with `v`; they are
    invalid according to :pep-reference:`0440` (thanks @virtuald &
    @stevejefferiesIDBS).
  - Support versions with epochs separated by `!` like `package-1!1.1.0`.
  - #102: FIX regression on uploading packages with `+` char in their version
    caused by recent bottle-upgrade.
- #103: Minor doc fixes (thanks @MichaelSchneeberger).


1.1.9 (2015-12-21)
------------------
"Ssss-elections" bug-fix & maintenance release.

- Upgrade bottle 1.11.6-->1.13-dev.

  - Fixes `MAX_PARAM` limiting dependencies(#82)

- Rework main startup and standalone:

  - New standalone generation based on ZIPed wheel archive.
  - Replace all sys.module mechanics with relative imports.
  - Fix gevent monkeypatching (#49).
  - Simplify definition of config-options on startup.
  - TODO: Move startup-options validations out of `main()` and
    into `pypiserver.core`
    package, to validate also start-up from API-clients.

- #53: Like PyPI, HREF-links now contain package's md5-hashes in their fragment.
  Add `--hash_algo` cmd-line option to turn-off or specify other *hashlib*
  message-digest algorithms (e.g. `sha256` is a safer choice, set it to `off`
  to avoid any performance penalty if hosting a lot of packages).

- #97: Add `--auther` non cmd-line startup-option to allow for alternative
  authentication methods (non HtPasswdFile-based one) to be defined by
  API-clients (thanks @Tythos).

- #91: Attempt to fix register http failures (thanks to @Tythos and @petri).

  - Test actual clients (ie `pip`, `Twine`, `setuptools`).
  - Test spurious `setuptools` failures.
  - NOT FIXED!  Still getting spurious failures.

- Various other fixes:

  - #96: Fix program's requirement (i.e. add passlib as extra-requirement).
    provide requirements files also for developers.
  - logging: Send also bottle `_stderr` to logger; fix logger names.
  - #95: Add missing loop-terminators in bottle-templates (thanks to @bmflynn).



1.1.8 (2015-09-15)
------------------
"Finikounda" release.

- Allow un-authenticated uploads (no htpasswd file) (#55).
- Fixes on package-name handling (#85 and #88, #89).
- Respect logging cmd-line options (#81).
- Add TCs for standalone script and other build-issues (#92)
- See milestone:M1.1.8 on github for all fixes included.


1.1.7 (2015-03-8)
------------------
1st release under cooperative ownership:

- #65, #66: Improve Auth for private repos by supporting i
  password protected package listings and downloads,
  in addition to uploads (use the -a, --authenticate option
  to specify which to protect).
- #67: Add cache-control http-header, reqed by pip.
- #56, #70: Ignore non-packages when serving.
- #58, #62: Log all http-requests.
- #61: Possible to change welcome-msg.
- #77, #78: Avoid XSS by generating web-content with SimpleTemplate
  instead of python's string-substs.
- #38, #79: Instruct to use --extra-index-url for misspelled dependencies to work,
  reorganize README instructions.


1.1.6 (2014-03-05)
------------------
- remove --index-url cli parameter introduced in 1.1.5

1.1.5 (2014-01-20)
------------------
- only list devpi-server and proxypypi as alternatives
- fix wheel file handling for certain wheels
- serve wheel files as application/octet-stream
- make pypiserver executable from wheel file
- build universal wheel
- remove scripts subdirectory
- add --index-url cli parameter

1.1.4 (2014-01-03)
------------------
- make pypiserver compatible with pip 1.5
  (https://github.com/pypiserver/pypiserver/pull/42)

1.1.3 (2013-07-22)
------------------
- make guessing of package name and version more robust

1.1.2 (2013-06-22)
------------------
- fix "pypi-server -U" stable/unstable detection, i.e. do not
  accidentally update to unstable packages

1.1.1 (2013-05-29)
------------------
- add 'overwrite' option to allow overwriting existing package
  files (default: false)

pypiserver/__main__.py
#! /usr/bin/env python
"""
.. NOTE:: To the developer:
    This module is moved to the root of the standalone zip-archive,
    to be used as its entry-point. Therefore DO NOT import relative.
"""
from __future__ import print_function

import os
import sys
import getopt
import re
import logging
import warnings
import textwrap
from pypiserver import ConfigError

warnings.filterwarnings("ignore", "Python 2.5 support may be dropped in future versions of Bottle")
log = logging.getLogger('pypiserver.main')


def init_logging(level=None, frmt=None, filename=None):
    logging.basicConfig(level=level, format=frmt)
    rlog = logging.getLogger()
    rlog.setLevel(level)
    if filename:
        rlog.addHandler(logging.FileHandler(filename))


def usage():
    return textwrap.dedent("""\
  pypi-server [OPTIONS] [PACKAGES_DIRECTORY...]
    start PyPI compatible package server serving packages from
    PACKAGES_DIRECTORY. If PACKAGES_DIRECTORY is not given on the
    command line, it uses the default ~/packages.  pypiserver scans this
    directory recursively for packages. It skips packages and
    directories starting with a dot. Multiple package directories can be
    specified.

  pypi-server understands the following options:

    -p, --port PORT
      listen on port PORT (default: 8080)

    -i, --interface INTERFACE
      listen on interface INTERFACE (default: 0.0.0.0, any interface)

    -a, --authenticate (UPDATE|download|list), ...
      comma-separated list of (case-insensitive) actions to authenticate
      Use '.' or '' for empty. Requires to have set the password (-P option).
      For example to password-protect package downloads (in addition to uploads)
      while leaving listings public, give:
        -P foo/htpasswd.txt  -a update,download
      To drop all authentications, use:
        -P .  -a .
      Note that when uploads are not protected, the `register` command
      is not necessary, but `~/.pypirc` still need username and password fields,
      even if bogus.
      By default, only 'update' is password-protected.

    -P, --passwords PASSWORD_FILE
      use apache htpasswd file PASSWORD_FILE to set usernames & passwords when
      authenticating certain actions (see -a option).
      If you want to allow un-authorized access, set this option and -a
      explicitly to empty (either '.' or'').

    --disable-fallback
      disable redirect to real PyPI index for packages not found in the
      local index

    --fallback-url FALLBACK_URL
      for packages not found in the local index, this URL will be used to
      redirect to (default: http://pypi.python.org/simple)

    --server METHOD
      use METHOD to run the server. Valid values include paste,
      cherrypy, twisted, gunicorn, gevent, wsgiref, auto. The
      default is to use "auto" which chooses one of paste, cherrypy,
      twisted or wsgiref.

    -r, --root PACKAGES_DIRECTORY
      [deprecated] serve packages from PACKAGES_DIRECTORY

    -o, --overwrite
      allow overwriting existing package files

    --hash-algo ALGO
      any `hashlib` available algo used as fragments on package links.
      Set one of (0, no, off, false) to disabled it. (default: md5)

    --welcome HTML_FILE
      uses the ASCII contents of HTML_FILE as welcome message response.

    -v
      enable verbose logging;  repeat for more verbosity.

    --log-file <FILE>
      write logging info into this FILE.

    --log-frmt <FILE>
      the logging format-string.  (see `logging.LogRecord` class from standard python library)
      [Default: %(asctime)s|%(levelname)s|%(thread)d|%(message)s]

    --log-req-frmt FORMAT
      a format-string selecting Http-Request properties to log; set to  '%s' to see them all.
      [Default: %(bottle.request)s]

    --log-res-frmt FORMAT
      a format-string selecting Http-Response properties to log; set to  '%s' to see them all.
      [Default: %(status)s]

    --log-err-frmt FORMAT
      a format-string selecting Http-Error properties to log; set to  '%s' to see them all.
      [Default: %(body)s: %(exception)s \n%(traceback)s]

    --cache-control AGE
      Add "Cache-Control: max-age=AGE, public" header to package downloads.
      Pip 6+ needs this for caching.


  pypi-server -h
  pypi-server --help
    show this help message

  pypi-server --version
    show pypi-server's version

  pypi-server -U [OPTIONS] [PACKAGES_DIRECTORY...]
    update packages in PACKAGES_DIRECTORY. This command searches
    pypi.python.org for updates and shows a pip command line which
    updates the package.

  The following additional options can be specified with -U:

    -x
      execute the pip commands instead of only showing them

    -d DOWNLOAD_DIRECTORY
      download package updates to this directory. The default is to use
      the directory which contains the latest version of the package to
      be updated.

    -u
      allow updating to unstable version (alpha, beta, rc, dev versions)

  Visit https://pypi.python.org/pypi/pypiserver for more information.
  """)


def main(argv=None):
    import pypiserver

    if argv is None:
        argv = sys.argv

    command = "serve"

    c = pypiserver.Configuration(**pypiserver.default_config())

    update_dry_run = True,
    update_directory = None,
    update_stable_only = True,

    try:
        opts, roots = getopt.getopt(argv[1:], "i:p:a:r:d:P:Uuvxoh", [
            "interface=",
            "passwords=",
            "authenticate=",
            "port=",
            "root=",
            "server=",
            "fallback-url=",
            "disable-fallback",
            "overwrite",
            "hash-algo=",
            "log-file=",
            "log-frmt=",
            "log-req-frmt=",
            "log-res-frmt=",
            "log-err-frmt=",
            "welcome=",
            "cache-control=",
            "version",
            "help"
        ])
    except getopt.GetoptError:
        err = sys.exc_info()[1]
        sys.exit("usage error: %s" % (err,))

    for k, v in opts:
        if k in ("-p", "--port"):
            c.port = int(v)
        elif k in ("-a", "--authenticate"):
            c.authenticated = v
        elif k in ("-i", "--interface"):
            c.host = v
        elif k in ("-r", "--root"):
            roots.append(v)
        elif k == "--disable-fallback":
            c.redirect_to_fallback = False
        elif k == "--fallback-url":
            c.fallback_url = v
        elif k == "--server":
            c.server = v
        elif k == "--welcome":
            c.welcome_file = v
        elif k == "--version":
            print("pypiserver %s\n" % pypiserver.__version__)
            return
        elif k == "-U":
            command = "update"
        elif k == "-x":
            update_dry_run = False
        elif k == "-u":
            update_stable_only = False
        elif k == "-d":
            update_directory = v
        elif k in ("-P", "--passwords"):
            c.password_file = v
        elif k in ("-o", "--overwrite"):
            c.overwrite = True
        elif k in ("--hash-algo"):
            c.hash_algo = None if not pypiserver.str2bool(v, c.hash_algo) else v
        elif k == "--log-file":
            c.log_file = v
        elif k == "--log-frmt":
            c.log_frmt = v
        elif k == "--log-req-frmt":
            c.log_req_frmt = v
        elif k == "--log-res-frmt":
            c.log_res_frmt = v
        elif k == "--log-err-frmt":
            c.log_err_frmt = v
        elif k == "--cache-control":
            c.cache_control = v
        elif k == "-v":
            c.verbosity += 1
        elif k in ("-h", "--help"):
            print(usage())
            sys.exit(0)


    verbose_levels = [logging.WARNING, 
                      logging.INFO, 
                      logging.DEBUG, 
                      logging.NOTSET]
    c.log_level = list(zip(verbose_levels, range(c.verbosity)))[-1][0]
    init_logging(level=c.log_level, filename=c.log_file, frmt=c.log_frmt)

    c.root = [os.path.abspath(x) for x in roots]

    if command == "update":
        from pypiserver.manage import update_all_packages
        update_all_packages(c.root, update_directory,
                dry_run=update_dry_run, stable_only=update_stable_only)
        return

    # Fixes #49:
    #    The gevent server adapter needs to patch some
    #    modules BEFORE importing bottle!
    if c.server and c.server.startswith('gevent'):
        import gevent.monkey  # @UnresolvedImport
        gevent.monkey.patch_all()

    try:
        app = pypiserver.app(**vars(c))
    except ConfigError:
        err = sys.exc_info()[1]
        sys.exit(err)

    from pypiserver import bottle
    bottle.debug(c.log_level <= logging.DEBUG)
    bottle.run(app=app, host=c.host, port=c.port, server=c.server)


if __name__ == "__main__":
    main()
